src/stdlib/io.rs

```diff
@@ -69,4 +88,831 @@ pub fn load_io<'gc>(ctx: Context<'gc>) {
+                }
+                Err(e) => {
+                    stack.replace(
+                        ctx,
+                        (Value::Nil, e.to_string(), e.raw_os_error().unwrap_or(0)),
+                    );
+                    Ok(CallbackReturn::Return)
+                }
+            }
+        }),
+    );
+
+    file.set_field(
+        ctx,
+        "read",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            if stack.is_empty() {
+                return Err("bad argument #1 to 'read' (file expected)"
+                    .into_value(ctx)
+                    .into());
+            }
+
+            let Some(file) = stack.pop_front() else {
+                return Err("bad argument #1 to 'read' (file expected)"
+                    .into_value(ctx)
+                    .into());
+            };
+            let file = if let Value::UserData(file) = file {
+                if let Ok(file) = file.downcast_static::<IoFile>() {
+                    file
+                } else {
+                    return Err("bad argument #1 to 'read' (file expected)"
+                        .into_value(ctx)
+                        .into());
+                }
+            } else {
+                return Err("bad argument #1 to 'read' (file expected)"
+                    .into_value(ctx)
+                    .into());
+            };
+
+            let formats = stack
+                .into_iter()
+                .enumerate()
+                .map(|(n, value)| {
+                    let Some(format) = value.into_string(ctx) else {
+                        return Err(format!(
+                            "bad argument #{} to 'read' (string expected, got {})",
+                            n + 1,
+                            value.type_name()
+                        )
+                        .into_value(ctx)
+                        .into());
+                    };
+                    match format.to_str() {
+                        Ok(s) => Ok(s),
+                        Err(err) => Err(err.to_string().into_value(ctx).into()),
+                    }
+                })
+                .collect::<Result<Vec<_>, Error<'_>>>()?;
+
+            stack.clear();
+            for format in formats {
+                match file.read_with_format(ctx, format)? {
+                    Some(value) => {
+                        stack.into_back(ctx, value);
+                    }
+                    None => {
+                        stack.into_back(ctx, Value::Nil);
+                    }
+                }
+            }
+
+            Ok(CallbackReturn::Return)
+        }),
+    );
+
+    file.set_field(
+        ctx,
+        "lines",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            #[derive(Collect, Clone)]
+            #[collect(require_static)]
+            struct Lines {
+                position: RefCell<usize>,
+                file: IoFile,
+                formats: Vec<std::string::String>,
```

Comment:
The formats should be parsed and stored as an enum, so that the rest of the implementation doesn't have to deal with possibly invalid format modes.